test/docfx.Test/lib/GitHubUserCacheTest.cs

```diff
@@ -0,0 +1,218 @@
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                0,
+                0);
+        }
+
+        [Fact]
+        public void GetUserByCommmitWithNewEmail()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByCommit("alice_new@contoso.com", "owner", "name", "1"),
+                "{}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\",\"alice_new@contoso.com\"]}]}",
+                1,
+                1);
+        }
+
+        [Fact]
+        public void GetUserByCommmitWithNewEmailCanCompleteCache()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByCommit("alice_new@contoso.com", "owner", "name", "1"),
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\",\"alice_new@contoso.com\"]}]}",
+                0,
+                1);
+        }
+
+        [Fact]
+        public void GetUserByInvalidLogin()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByLogin("invalid"),
+                "{}",
+                "{\"users\":[{\"login\":\"invalid\",\"emails\":[]}]}",
+                1,
+                0); ;
+        }
+        [Fact]
+        public void GetUserByInvalidLoginFromCache()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByLogin("invalid"),
+                "{\"users\":[{\"login\":\"invalid\",\"emails\":[]}]}",
+                "{\"users\":[{\"login\":\"invalid\",\"emails\":[]}]}",
+                0,
+                0); ;
+        }
+
+        private async void VerifyCore(Action<GitHubUserCache> test, string inputCache, string expectedOutputCache, int expectedGetUserByLoginCall, int expectedGetLoginByCommitCall)
+        {
+            // Arrange
+            var docsetPath = Path.Combine("github-user-cache-test", Path.GetRandomFileName());
+            if (Directory.Exists(docsetPath))
+                Directory.Delete(docsetPath, true);
+            Directory.CreateDirectory(docsetPath);
+            var docfxYmlPath = Path.Combine(docsetPath, "docfx.yml");
+            var gitHubUserCachePath = Path.Combine(docsetPath, "cache.json");
+            File.WriteAllText(docfxYmlPath, @"gitHub:
+  userCache: cache.json");
+            File.WriteAllText(gitHubUserCachePath, inputCache);
+
+            var (_, config) = Config.Load(docsetPath, new CommandLineOptions());
+            var options = new CommandLineOptions();
+            var docset = new Docset(new Context(new Report(), "_site"), docsetPath, config, options);
+            var cache = await GitHubUserCache.Create(docset, null);
```

Comment:
We can create a `GitHubUserCache(string cachePath, int expirationInHours)` constructor and a `IEnumerable<GitHubUser> Users` property for testing purpose, so you don't need to read/write files